Batteries/Data/Fin/Coding.lean

```diff
@@ -0,0 +1,539 @@
+/-
+Copyright (c) 2024 François G. Dorais. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: François G. Dorais
+-/
+import Batteries.Data.Char
+import Batteries.Data.Fin.Lemmas
+import Batteries.Tactic.Basic
+import Batteries.Tactic.Trans
+import Batteries.Tactic.Lint
+
+/-! # Coding recipes for `Fin` types
+
+This module collects encoding/decoding bijections to represent basic type constructors of `Fin`
+types as `Fin` types. These functions implement the isomorphism `Option (Fin n) ≃ Fin (n+1)`,
+`Fin m ⊕ Fin n ≃ Fin (m + n)`, `Fin m × Fin n ≃ Fin (m * n)`, ..., including, dependent sums,
+dependent products, decidable subtypes and decidable quotients.
+-/
+
+namespace Fin
+
+/-- Encode a unit value as a `Fin` type. -/
+@[nolint unusedArguments]
+def encodePUnit : PUnit → Fin 1
+  | .unit => 0
+
+/-- Decode a unit value as a `Fin` type. -/
+@[pp_nodot] def decodePUnit : Fin 1 → PUnit
+  | 0 => .unit
+
+/-- Encode a unit value as a `Fin` type. -/
+abbrev encodeUnit : Unit → Fin 1 := encodePUnit
+
+/-- Decode a unit value as a `Fin` type. -/
+@[pp_nodot] abbrev decodeUnit : Fin 1 → Unit := decodePUnit
+
+@[simp] theorem encodePUnit_decodePUnit : (x : Fin 1) → encodePUnit (decodePUnit x) = x
+  | 0 => rfl
+
+@[simp] theorem decodePUnit_encodePUnit : (x : PUnit) → decodePUnit (encodePUnit x) = x
+  | .unit => rfl
+
+/-- Encode a boolean value as a `Fin` type. -/
+def encodeBool : Bool → Fin 2
+  | false => 0
+  | true => 1
+
+/-- Decode a boolean value as a `Fin` type. -/
+@[pp_nodot] def decodeBool : Fin 2 → Bool
+  | 0 => false
+  | 1 => true
+
+@[simp] theorem encodeBool_decodeBool : (x : Fin 2) → encodeBool (decodeBool x) = x
+  | 0 => rfl
+  | 1 => rfl
+
+@[simp] theorem decodeBool_encodeBool : (x : Bool) → decodeBool (encodeBool x) = x
+  | false => rfl
+  | true => rfl
+
+/-- Encode a character value as a `Fin` type.
+
+Valid character code points range from 0 to 1114112 (U+10FFFF) excluding the surrogate range from
+55296 (U+D800) to 57343 (U+DFFF). This results in 1112064 valid code points.
+(See [unicode scalar value](https://www.unicode.org/glossary/#unicode_scalar_value).)
+-/
+def encodeChar (c : Char) : Fin 1112064 :=
+  have : c.toNat < 1114112 :=
```

Comment:
This looks like it would be a good lemma in the `Char` namespace.

Reply:
Maybe even with a `def` for the magic constant.

Reply:
Done. I also fixed the constants for `Char.all`/`Char.any` at the same time.